openapi.yaml

```diff
@@ -0,0 +1,484 @@
+        * If `gitRepo` is already set on creation, no `<GitRepoSpec>` will be added.
+      tags:
+        - tenant
+      requestBody:
+        required: true
+        $ref: '#/components/requestBodies/CreateTenant'
+      responses:
+        '201':
+          description: Tenant created
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TenantId'
+              example:
+                id: aezoo6
+        '400':
+          description: Tenant cannot be created
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Reason'
+        '405':
+          description: Tenant already exists
+  /tenants/{tenantId}:
+    get:
+      summary: Returns all values of a tenant
+      tags:
+        - tenant
+      parameters:
+        - $ref: '#/components/parameters/TenantId'
+      responses:
+        '200':
+          description: Tenant found
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Tenant'
+        '404':
+          description: A tenant with the specified id was not found.
+    patch:
+      summary: Updates a tenant
+      tags:
+        - tenant
+      parameters:
+        - $ref: '#/components/parameters/TenantId'
+      requestBody:
+        description: Update tenant with properties to be changed (RFC 7396)
+        required: true
+        content:
+          application/merge-patch+json:
+            schema:
+              allOf:
+                - $ref: '#/components/schemas/Tenant'
+                - type: object
+                  properties:
+                    displayName:
+                      nullable: true
+      responses:
+        '204':
+          description: Tenant updated
+        '403':
+          description: Tenant update forbidden
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Reason'
+    delete:
+      summary: Deletes a tenant
+      tags:
+        - tenant
+      responses:
+        '204':
+          description: Tenant deleted
+        '403':
+          description: Tenant deletion forbidden
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Reason'
+  /clusters:
+    get:
+      summary: Returns a list of clusters.
+      description: List of clusters available in the API
+      tags:
+        - cluster
+      parameters:
+        - in: query
+          name: tenant
+          schema:
+            type: string
+          description: Filter clusters by tenant id
+          example: aezoo6
+      responses:
+        '200':
+          description: Clusters found
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Cluster'
+        '404':
```

Comment:
Is this really 404 or just an empty array?

Reply:
Unsure. We could either

1. send an empty array with a 404
2. send an empty array with a 200
3. just send a 404 and leave the handling up to the consumer

From a HTTP point of view I would either choose 1 or 3. 3 seems a bit more the way I see handling of such cases, but could make the client implementation a bit more complex. What would you prefer?

Reply:
I was more thinking of 2. (empty & 200) because it's not really an error, right? Also if we introduce more query options, it might very well be that your query doesn't match any cluster.

Reply:
Makes sense! I changed the behavior this way.